cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/engine/sql/dataset/SQLDataset.java

```diff
@@ -26,4 +26,12 @@ public interface SQLDataset extends SQLDatasetDescription {
    */
   long getNumRows();
 
+  /**
+   * Method to determine if this SQL dataset is a valid dataset that is correctly pushed into the SQL engine
+   * @return boolean specifying if this stage is pushed into the SQL engine
+   */
+  default boolean isValid() {
```

Comment:
What happens if it's not in general? I mean, overall we should not introduce it just for reading as it means we need to handle it for all cases

Reply:
Also I feel it's often too late to change gears when SQLDataset is returned. We generally have a "planning" phase and "execution" phase and any decisions should be done during planning phase otherwise our decision strategy would have hard time making decisions

Reply:
Having this method allows us to have the Read operation return a `SQLEngineJob<SQLDataset>` response.

If the SQLDataset entity is valid, we know it's a dataset that represent records that have been written into the SQL engine and we can operate on.

If the SQLDataset entity is **not** valid, we know that something caused the read operation to fail. Thus, we fall back to the default workflow for the source.

Reply:
But you are adding it to the generic SQLDataset interface. It means anyone using it should support invalid dataset, not only  in Read

Reply:
But, to your point, we can make the `read` method in the adapter return the instance of `SQLReadResult` and implement the method to verify validity of the result in the `SQLReadResult` class. This way we don't have to change the SQLDataset method signature.

Reply:
Overall this question is secondary to "do we need it". Let's decide that first and come back here if needed